kdc/bx509d.c

```diff
@@ -1843,6 +1939,189 @@ get_tgt(struct bx509_request_desc *r)
+        if (ret)
+            return bad_500(r, ret, "Out of resources");
+    }
+    heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                     "requested_krb5PrincipalName", "%s", val);
+    ret = hx509_request_add_pkinit(r->context->hx509ctx, r->req,
+                                   val);
+    if (ret)
+        return bad_403(r, ret, "Not authorized to requested TGT");
+    return MHD_YES;
+}
+
+/* For each requested principal, produce a ccache */
+static heim_mhd_result
+get_tgts_param_execute_cb(void *d,
+                          enum MHD_ValueKind kind,
+                          const char *key,
+                          const char *val)
+{
+    struct bx509_request_desc *r = d;
+    heim_mhd_result res = MHD_YES;
+
+    if (r->error_code == 0 &&
+        strcmp(key, "cname") == 0 && val) {
+        /* Handled upstairs */
+        r->for_cname = val;
+        r->error_code = k5_get_creds(r, K5_CREDS_EPHEMERAL);
+        if (r->error_code == 0)
+            res = get_tgts_accumulate_ccache(r);
+    } else {
+        /* Handled upstairs */
+    }
+    return res;
+}
+
+/*
+ * Implements /get-tgts end-point.
+ *
+ * Query parameters:
+ *
+ *  - cname=<name> (client principal name, if not the same as the authenticated
+ *                  name, then this will be impersonated if allowed; may be
+ *                  given multiple times)
+ */
+static krb5_error_code
+get_tgts(struct bx509_request_desc *r)
+{
+    krb5_error_code ret;
+    krb5_principal p = NULL;
+    size_t bodylen;
+    void *body;
+
+    /* Prep to authorize */
+    ret = krb5_parse_name(r->context, r->cname, &p);
+    if (ret)
+        return bad_403(r, ret, "Could not parse caller principal name");
+    if (ret == 0) {
+        /* Extract q-params other than `cname' */
+        r->error_code = 0;
+        (void) MHD_get_connection_values(r->connection, MHD_GET_ARGUMENT_KIND,
+                                         get_tgt_param_cb, r);
+        ret = r->error_code;
+    }
+    if (ret == 0) {
+        /* Authorize requested client principal names (calls bad_req()) */
+        r->error_code = 0;
+        (void) MHD_get_connection_values(r->connection, MHD_GET_ARGUMENT_KIND,
+                                         get_tgts_param_authorize_cb, r);
+        ret = r->error_code;
+        if (ret == 0) {
+            ret = kdc_authorize_csr(r->context, "get-tgt", r->req, p);
+            if (ret) {
+                krb5_free_principal(r->context, p);
+                return bad_403(r, ret, "Permission denied");
+            }
+        }
+        hx509_request_free(&r->req);
+    }
+    if (ret == 0) {
+        /* get_tgts_param_execute_cb() calls bad_req() */
+        r->error_code = 0;
+        (void) MHD_get_connection_values(r->connection, MHD_GET_ARGUMENT_KIND,
+                                         get_tgts_param_execute_cb, r);
+        ret = r->error_code;
+    }
+    krb5_free_principal(r->context, p);
+    if (ret)
+        return ret;
```

Comment:
Return some `bad_404()` or something.

Reply:
No, just never fail except for 401/403/5xx reasons.  For individual principals, write a JSON text bearing an error code and an error message in the body.

Reply:
Fixed.  Though if authorization fails, we do fail the whole request.  Making it so we don't do that will require a fair bit more work.